File: src/com/example/javas/easy/E016_SymmetricTree.java
/*
EASY PROBLEM (Simplified for 15 min):
Symmetric Tree - Check if Binary Tree is Mirror of Itself

REQUIREMENTS:
- Given the root of a binary tree
- Return true if the tree is symmetric (mirror of itself around center)
- Check both structure and values

APPROACH:
- Use recursive helper to compare left and right subtrees
- Left subtree's left should match right subtree's right
- Left subtree's right should match right subtree's left

TIME COMPLEXITY: O(n) where n is number of nodes
SPACE COMPLEXITY: O(h) where h is height (recursion stack)
*/

package com.example.javas.easy;

public class E016_SymmetricTree {
    
    // TreeNode definition
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        
        TreeNode(int val) {
            this.val = val;
        }
    }
    
    public static boolean isSymmetric(TreeNode root) {
        // TODO: Implement symmetric checker
        // 1. If root is null, return true
        // 2. Call helper to compare left and right subtrees
        
        return false;
    }
    
    private static boolean isMirror(TreeNode left, TreeNode right) {
        // TODO: Implement mirror checker
        // 1. If both null, return true
        // 2. If one null, return false
        // 3. If values different, return false
        // 4. Recursively check:
        //    - left.left vs right.right AND
        //    - left.right vs right.left
        
        return false;
    }
    
    // TEST CASES
    public static void main(String[] args) {
        // Symmetric tree:    1
        //                   / \
        //                  2   2
        //                 / \ / \
        //                3  4 4  3
        TreeNode root1 = new TreeNode(1);
        root1.left = new TreeNode(2);
        root1.right = new TreeNode(2);
        root1.left.left = new TreeNode(3);
        root1.left.right = new TreeNode(4);
        root1.right.left = new TreeNode(4);
        root1.right.right = new TreeNode(3);
        System.out.println("Symmetric tree [1,2,2,3,4,4,3]: " + isSymmetric(root1));
        
        // Not symmetric:  1
        //                / \
        //               2   2
        //                \   \
        //                 3   3
        TreeNode root2 = new TreeNode(1);
        root2.left = new TreeNode(2);
        root2.right = new TreeNode(2);
        root2.left.right = new TreeNode(3);
        root2.right.right = new TreeNode(3);
        System.out.println("Symmetric tree [1,2,2,null,3,null,3]: " + isSymmetric(root2));
        
        // Single node
        TreeNode root3 = new TreeNode(1);
        System.out.println("Symmetric single node: " + isSymmetric(root3));
        
        // Null tree
        System.out.println("Symmetric null tree: " + isSymmetric(null));
    }
}

File: pythonX/easy/e016_symmetric_tree.py
"""
E016: Symmetric Tree - Check if Binary Tree is Mirror of Itself

Summary:
Symmetric Tree

Requirements:
- Given the root of a binary tree
- Return true if the tree is symmetric (mirror of itself around center)
- Check both structure and values

Approach:
- Use recursive helper to compare left and right subtrees
- Left subtree's left should match right subtree's right
- Left subtree's right should match right subtree's left

Big O Analysis:
Time Complexity: O(n) where n is number of nodes
Space Complexity: O(h) where h is height (recursion stack)
"""

from typing import Optional


class TreeNode:
    """Definition for a binary tree node."""
    def __init__(self, val: int = 0, left: 'Optional[TreeNode]' = None, right: 'Optional[TreeNode]' = None):
        self.val = val
        self.left = left
        self.right = right


def is_symmetric(root: Optional[TreeNode]) -> bool:
    """
    TODO: Implement symmetric checker
    1. If root is None, return True
    2. Call helper to compare left and right subtrees
    """
    pass


def is_mirror(left: Optional[TreeNode], right: Optional[TreeNode]) -> bool:
    """
    TODO: Implement mirror checker
    1. If both None, return True
    2. If one None, return False
    3. If values different, return False
    4. Recursively check:
       - left.left vs right.right AND
       - left.right vs right.left
    """
    pass


if __name__ == "__main__":
    # Test Cases
    
    # Symmetric tree:    1
    #                   / \
    #                  2   2
    #                 / \ / \
    #                3  4 4  3
    root1 = TreeNode(1)
    root1.left = TreeNode(2)
    root1.right = TreeNode(2)
    root1.left.left = TreeNode(3)
    root1.left.right = TreeNode(4)
    root1.right.left = TreeNode(4)
    root1.right.right = TreeNode(3)
    print(f"Symmetric tree [1,2,2,3,4,4,3]: {is_symmetric(root1)}")
    
    # Not symmetric:  1
    #                / \
    #               2   2
    #                \   \
    #                 3   3
    root2 = TreeNode(1)
    root2.left = TreeNode(2)
    root2.right = TreeNode(2)
    root2.left.right = TreeNode(3)
    root2.right.right = TreeNode(3)
    print(f"Symmetric tree [1,2,2,null,3,null,3]: {is_symmetric(root2)}")
    
    # Single node
    root3 = TreeNode(1)
    print(f"Symmetric single node: {is_symmetric(root3)}")
    
    # None tree
    print(f"Symmetric null tree: {is_symmetric(None)}")
